llvm/lib/CodeGen/CodeGenPrepare.cpp

```diff
@@ -2621,8 +2622,36 @@ bool CodeGenPrepare::dupRetToEnableTailCallOpts(BasicBlock *BB,
       // Make sure the phi value is indeed produced by the tail call.
       if (CI && CI->hasOneUse() && CI->getParent() == PredBB &&
           TLI->mayBeEmittedAsTailCall(CI) &&
-          attributesPermitTailCall(F, CI, RetI, *TLI))
+          attributesPermitTailCall(F, CI, RetI, *TLI)) {
         TailCallBBs.push_back(PredBB);
+      } else {
+        /// Consider the cases in which the phi value is indirectly produced by
+        /// the tail call, for example when encountering memset(), memmove(),
+        /// whose return value may have been optimized out.
+        /// @code
+        /// bb0:
+        ///   tail call void @llvm.memset.p0.i64()
+        ///   br label %return
+        /// @endcode
+        if (PredBB && PredBB->getSingleSuccessor() == BB)
+          CI = dyn_cast_or_null<CallInst>(
+              PredBB->getTerminator()->getPrevNonDebugInstruction(true));
+
+        // If we return void from the call site, it must be an intrinsic.
+        if (CI && CI->getType()->isVoidTy() && !isa<IntrinsicInst>(CI))
```

Comment:
Is there a need to check which arg is fed to the phi and returned? The arg needs to be the real return value of the lowered function call?

Reply:
My question is if this case https://godbolt.org/z/jndea9van is properly detected and skipped.

Reply:
Oh sure, thanks, added an extra check skipping such cases and this one (https://godbolt.org/z/3fce9qrvz) as well – we need the argument fed to the phi to be the first one.

Reply:
Reviewed test too.